pkg/filesystem/referer_adaptor.go

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package filesystem
+
+import (
+	"context"
+	"fmt"
+
+	snpkg "github.com/containerd/containerd/pkg/snapshotters"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+)
+
+func (fs *Filesystem) CheckReferrer(ctx context.Context, labels map[string]string) bool {
+	if !fs.ReferrerDetectEnabled() {
+		return false
+	}
+
+	ref, ok := labels[snpkg.TargetRefLabel]
+	if !ok {
+		return false
+	}
+
+	manifestDigest := digest.Digest(labels[snpkg.TargetManifestDigestLabel])
+	if manifestDigest.Validate() != nil {
+		return false
+	}
+
+	if _, err := fs.referrerMgr.CheckReferrer(ctx, ref, manifestDigest); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (fs *Filesystem) TryFetchMetadata(ctx context.Context, labels map[string]string, metadataPath string) error {
+	ref, ok := labels[snpkg.TargetRefLabel]
+	if !ok {
+		return fmt.Errorf("invalid label %s=%s", snpkg.TargetRefLabel, ref)
```

Comment:
`ref` must be its zero value. The error message seems meaningless

Reply:
Fixed